drv/gimlet-seq-server/src/vcore.rs

```diff
@@ -0,0 +1,137 @@
+/// of these READ_VIN measurements, along with timestamps before and after the
+/// operations, and put them all in a ring buffer.  Note that we don't clear
+/// faults after this condition; we will wait until the machine next makes an
+/// A2 to A0 transition to clear faults.
+///
+use drv_i2c_api::{I2cDevice, ResponseCode};
+use drv_i2c_devices::raa229618::Raa229618;
+use drv_stm32xx_sys_api as sys_api;
+use ringbuf::*;
+use sys_api::IrqControl;
+use userlib::*;
+
+pub struct VCore {
+    device: Raa229618,
+    sys: sys_api::Sys,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    Notified,
+    Fault,
+    Start(u64),
+    Reading(units::Volts),
+    Error(ResponseCode),
+    Done(u64),
+    None,
+}
+
+ringbuf!(Trace, 120, Trace::None);
+
+///
+/// We are going to set our input undervoltage warn limit to be 11.75 volts.
+/// Note that we will not fault if VIN goes below this (that is, we will not
+/// lose POWER_GOOD), but the part will indicate an input fault and pull
+/// PWR_CONT1_VCORE_TO_SP_ALERT_L low.
+///
+const VCORE_UV_WARN_LIMIT: units::Volts = units::Volts(11.75);
+
+///
+/// We want to collect enough samples (at ~900µs per sample) to adequately
+/// cover any anticipated dip.  We have seen these have an ~11ms total width
+/// in the wild, so we give ourselves plenty of margin here and get ~45ms
+/// of data.
+///
+const VCORE_NSAMPLES: usize = 50;
+
+const VCORE_TO_SP_ALERT_L: sys_api::PinSet = sys_api::Port::I.pin(14);
+const VCORE_TO_SP_ALERT_PULL: sys_api::Pull = sys_api::Pull::None;
+
+cfg_if::cfg_if! {
+    if #[cfg(not(any(
+        target_board = "gimlet-b",
+        target_board = "gimlet-c",
+        target_board = "gimlet-d",
+        target_board = "gimlet-e",
+        target_board = "gimlet-f",
+    )))] {
+        compile_error!("RAA229618 VIN monitoring unsupported for this board");
+    }
+}
+
+impl VCore {
+    pub fn new(sys: &sys_api::Sys, device: &I2cDevice, rail: u8) -> Self {
+        Self {
+            device: Raa229618::new(device, rail),
+            sys: sys.clone(),
+        }
+    }
+
+    pub fn mask(&self) -> u32 {
+        crate::notifications::VCORE_MASK
+    }
+
+    pub fn initialize_uv_warning(&self) -> Result<(), ResponseCode> {
+        let sys = &self.sys;
+
+        // Set our warn limit
+        self.device.set_vin_uv_warn_limit(VCORE_UV_WARN_LIMIT)?;
+
+        // Clear our faults
+        self.device.clear_faults()?;
+
+        // Set our alert line to be an input
+        sys.gpio_configure_input(VCORE_TO_SP_ALERT_L, VCORE_TO_SP_ALERT_PULL);
+        sys.gpio_irq_configure(self.mask(), sys_api::Edge::Falling);
+
+        // Enable the interrupt!
+        let _ = self.sys.gpio_irq_control(self.mask(), IrqControl::Enable);
+
+        Ok(())
+    }
+
+    pub fn handle_notification(&self) {
+        let faulted = self.sys.gpio_read(VCORE_TO_SP_ALERT_L) == 0;
+
+        ringbuf_entry!(Trace::Notified);
+
+        if faulted {
+            ringbuf_entry!(Trace::Fault);
```

Comment:
Nitpicky, take it or leave it: it seems like we could, potentially, consume one fewer ringbuf slot if we combined these entries into a `Trace::Notified { fault: bool }`? This wouldn't make the size of the entry type any bigger, since we already record variants with data...

Reply:
Probably leaving that one. ;)